### PR TITLE
Refactor: fault types

### LINKED ISSUE #332: Refactor fault types
## User stories
As a developer I want to minimize code duplication in order to eliminate waste.
To achieve that, I want to move the `__init__` functions to the fault superclass. All but one fault (ScheduleBlockedFault) need the same attributes.

## Dependencies
Depends on: 
 - #316 
 - #317 

## Acceptance Criteria
- [ ] add new attributes to the Fault class
- [ ] expand the `__init__` method
- [ ] remove now obsolete attributes from the specific fault classes
- [ ] remove `__init__` methods from the specific fault classes

### PR DESCRIPTION
**Pull request by instant-sky**
Fixes #332 

Description (what might a PO or reviewer want to know?)

## PR checklist

- [ ] Acceptance criteria fulfilled
- [ ] Additional features are tested
- [ ] Docs (code, wiki & diagrams) updated
- [ ] Breaking changes anounced
- [ ] Added classes that inherit from `BaseModel` to `src.constants.tables`
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [ ] Tests and linter have passed AFTER local merge
- [ ] Another dev reviewed and approved

### CODE REVIEW

File: src/fault_injector/fault_types/fault.py [hunk 1]
```diff
@@ -46,3 +61,43 @@ def next_tick(self, tick: int):
             self.inject_fault(tick=tick)
         elif tick == self.configuration.end_tick:
             self.resolve_fault(tick=tick)
+
+
+class TrainMixIn:
```
Comment (antonykamp) [inline src/fault_injector/fault_types/fault.py:66]: Very cool idea with `TrainMixIn` and `TrackMixIn` 😮 I like it 💚

File: src/fault_injector/fault_types/fault.py [hunk 2]
```diff
@@ -46,3 +61,43 @@ def next_tick(self, tick: int):
...
+        """returns the train in which the requested fault
+
+        :param wrapper: wrapper component
+        :type wrapper: SimulationObjectUpdatingComponent
+        :param affected_element_id: id of the train in which the fault should be injected into
+        :type affected_element_id: str
+        :return: the train with the requested id
+        :rtype: Train
+        """
+        return [
```
Comment (antonykamp) [inline src/fault_injector/fault_types/fault.py:83]: Don't forget to check, if the array contains at least one element (or exactly one) 😉

File: src/fault_injector/fault_types/fault.py [hunk 3]
```diff
@@ -46,3 +61,43 @@ def next_tick(self, tick: int):
...
+        """returns the track in which the requested fault
+
+        :param wrapper: wrapper component
+        :type wrapper: SimulationObjectUpdatingComponent
+        :param affected_element_id: id of the track in which the fault should be injected into
+        :type affected_element_id: str
+        :return: the track with the requested id
+        :rtype: Track
+        """
+        return [
```
Comment (antonykamp) [inline src/fault_injector/fault_types/fault.py:107]: Don't forget to check, if you can find a track with this id 😉

File: src/fault_injector/fault_types/fault.py [hunk 4]
```diff
@@ -3,18 +3,33 @@
...
     logger: Logger
+    wrapper: SimulationObjectUpdatingComponent
+    interlocking: IInterlockingDisruptor
 
-    def __init__(self, configuration, logger: Logger):
+    def __init__(
+        self,
+        configuration,
+        logger: Logger,
+        wrapper: SimulationObjectUpdatingComponent,
```
Comment (antonykamp) [inline src/fault_injector/fault_types/fault.py]: I'm unsure if "wrapper" is the best name for this attribute 🙃 What do you think about `simulation_object_updater`?

File: src/fault_injector/fault_types/fault.py (hunk 1 above)
Comment (instant-sky) [reply src/fault_injector/fault_types/fault.py:66]: Credits for that go to @CR1337 😄

File: src/fault_injector/fault_types/fault.py (hunk 2 above)
Comment (instant-sky) [reply src/fault_injector/fault_types/fault.py:83]: Yes good catch! I will fix this in #353

File: src/fault_injector/fault_types/fault.py (hunk 4 above)
Comment (instant-sky) [reply src/fault_injector/fault_types/fault.py]: `wrapper` is closer to the component scheme I use for the other components, but you are right, `simulation_object_updater` is closer to the actual classes!

**Review by antonykamp — APPROVED** [review]
Great refactoring 🥳

**Pull request merged**